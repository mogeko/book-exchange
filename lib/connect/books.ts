import useSWR, { type SWRResponse } from "swr";
import handleQueryParam, { type QueryParamType } from "@/lib/utils/queryTools";

function useBooks(queryParam: QueryParamType<keyof BookTypes> = {}) {
  const query = queryParam
    ? `/books?${handleQueryParam(queryParam)}`
    : "/books";
  const { data, error }: SWRResponse<BookTypes[]> = useSWR(query);
  return {
    books: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export function useBook(
  id: string | number,
  queryParam: QueryParamType<keyof BookTypes> = {}
) {
  const query = queryParam
    ? `/books/${id}?${handleQueryParam(queryParam)}`
    : `/books/${id}`;
  const { data, error }: SWRResponse<BookTypes> = useSWR(query);
  return {
    book: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export interface BookTypes {
  author: string;
  cover: string;
  description: string;
  id: number;
  published: string;
  publisher: string;
  title: string;
}

export default useBooks;
