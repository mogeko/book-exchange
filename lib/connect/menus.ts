import useSWR, { type SWRResponse } from "swr";

const fetcher = (url: string) => fetch(url).then((r) => r.json());

function useMenus(url: string) {
  const { data, error }: SWRResponse<ResProps> = useSWR(url, fetcher);
  return {
    menus: data,
    isLoading: !error && !data,
    isError: error,
  };
}

interface ResProps {
  data: {
    href: string;
    name: string;
    badge?: string | number;
  }[];
}

export default useMenus;
