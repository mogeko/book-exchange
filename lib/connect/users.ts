import useSWR, { type SWRResponse } from "swr";
import handleQueryParam, { type QueryParamType } from "@/lib/utils/queryTools";

function useUsers(queryParam: QueryParamType<keyof UserType> = {}) {
  const query = queryParam
    ? `/users?${handleQueryParam(queryParam)}`
    : "/users";
  const { data, error }: SWRResponse<UserType[]> = useSWR(query);
  return {
    users: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export function useUser(
  id: string | number,
  queryParam: QueryParamType<keyof UserType> = {}
) {
  const query = queryParam
    ? `/users/${id}?${handleQueryParam(queryParam)}`
    : `/users/${id}`;
  const { data, error }: SWRResponse<UserType> = useSWR(query);
  return {
    user: data,
    isLoading: !error && !data,
    isError: error,
  };
}

export interface UserType {
  avatar: string;
  city: string;
  description: string;
  email: string;
  id: number;
  username: string;
}

export default useUsers;
