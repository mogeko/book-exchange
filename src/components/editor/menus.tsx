import classNames from "@/lib/utils/classNames";
import { BubbleMenu as Bubble, FloatingMenu as Floating } from "@tiptap/react";
import { FaLink } from "react-icons/fa";
import { useRef, useState, useEffect } from "react";
import { useCallback, useMemo } from "react";
import type { ButtonProps } from "@/components/editor/buttons";
import type { Editor } from "@tiptap/react";

export const BubbleMenu: React.FC<MenuProps> = ({ editor, buttons }) => {
  const [isLinkActive, setIsLinkActive] = useState(false);
  const [hasLinkInput, setHasLinkInput] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);
  const handleSubmit = useCallback(() => {
    if (inputRef.current?.value) {
      editor?.chain().focus().setLink({ href: inputRef.current?.value }).run();
    } else {
      editor?.chain().focus().unsetLink().run();
    }
    setIsLinkActive(!isLinkActive);
  }, [editor, isLinkActive, setIsLinkActive]);
  const LinkDialog = useMemo(() => {
    return (
      <div className="input-group">
        <input
          onChange={(e) => setHasLinkInput(e.target.value !== "")}
          defaultValue={editor?.getAttributes("link").href}
          className="input input-bordered input-xs"
          ref={inputRef}
          autoFocus
        />
        <button
          onClick={handleSubmit}
          className={classNames(
            hasLinkInput ? "btn-primary" : "",
            "btn btn-xs"
          )}
        >
          OK
        </button>
      </div>
    );
  }, [editor, hasLinkInput, handleSubmit]);
  const LinkButton = useMemo(() => {
    return (
      <button
        onClick={() => setIsLinkActive(!isLinkActive)}
        className={classNames(
          editor?.isActive("link") ? "btn-active" : "",
          "btn btn-xs"
        )}
      >
        <FaLink className="w-3 h-3" />
      </button>
    );
  }, [editor, isLinkActive, setIsLinkActive]);

  useEffect(() => {
    editor?.isFocused && setIsLinkActive(false);
  }, [editor?.isFocused, setIsLinkActive]);

  return (
    editor && (
      <Bubble
        className="btn-group"
        tippyOptions={{ duration: 100 }}
        editor={editor}
      >
        {!isLinkActive &&
          buttons.map((Button, i) => <Button key={i} editor={editor} />)}
        {!isLinkActive && LinkButton}
        {isLinkActive && LinkDialog}
      </Bubble>
    )
  );
};

export const FloatingMenu: React.FC<MenuProps> = ({ editor, buttons }) => {
  return (
    editor && (
      <Floating
        className="btn-group"
        tippyOptions={{ duration: 100 }}
        editor={editor}
      >
        {buttons.map((Button, i) => (
          <Button key={i} editor={editor} />
        ))}
      </Floating>
    )
  );
};

interface MenuProps {
  editor: Editor | null;
  buttons: React.FC<ButtonProps>[];
}
