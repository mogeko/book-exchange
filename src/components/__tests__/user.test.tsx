import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import User from "@/components/user";
import * as menus from "@/lib/connect/menus";

describe("User", () => {
  it("renders a user menu", () => {
    const { container } = render(<User />);
    expect(container.querySelector("a")?.textContent).toEqual(
      "Sign in / Sign up"
    );
  });

  it("snapshot a user menu", () => {
    const { container } = render(<User />);
    expect(container).toMatchSnapshot();
  });

  it("renders a user menu with a user", () => {
    render(<User username="testUser" />);
    expect(screen.getByText("testUser")).toBeInTheDocument();
    expect(screen.getByText("Profile")).toBeInTheDocument();
    expect(screen.getByText("Logout")).toBeInTheDocument();
  });

  it("snapshot a user menu with a user", () => {
    const { container } = render(<User username="testUser" />);
    expect(container).toMatchSnapshot();
  });

  it("renders a user menu for mobile", () => {
    const { container } = render(<User.Mobile />);
    expect(container.querySelector("svg")).toBeInTheDocument();
  });

  it("snapshot a user menu for mobile", () => {
    const { container } = render(<User.Mobile />);
    expect(container).toMatchSnapshot();
  });
});
