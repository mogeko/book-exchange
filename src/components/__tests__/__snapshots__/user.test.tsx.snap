// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User snapshot a user menu 1`] = `
<div>
  <a
    class="btn btn-link btn-xs normal-case"
  >
    Sign in / Sign up
  </a>
</div>
`;

exports[`User snapshot a user menu for mobile 1`] = `
<div>
  <div
    class="dropdown dropdown-end sm:hidden"
  >
    <label
      class="btn btn-ghost btn-circle avatar"
      tabindex="0"
    >
      <div
        class="w-6 rounded-full"
      >
        <svg
          aria-hidden="true"
          class="h-6 w-6"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 20 20"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M10 9a3 3 0 100-6 3 3 0 000 6zm-7 9a7 7 0 1114 0H3z"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </label>
    <ul
      class="menu menu-compact dropdown-content mt-3 p-2 shadow bg-base-200 rounded-box w-52"
      tabindex="0"
    >
      <li>
        <a
          href="/#"
        >
          Profile
        </a>
      </li>
      <li>
        <a
          href="/#"
        >
          Settings
        </a>
      </li>
      <li>
        <a
          href="/#"
        >
          Logout
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`User snapshot a user menu with a user 1`] = `
<div>
  <div
    class="dropdown dropdown-end"
  >
    <label
      class="btn btn-link btn-xs normal-case text-base-content"
      tabindex="0"
    >
      Hi!Â 
      <b>
        testUser
      </b>
    </label>
    <ul
      class="menu menu-compact dropdown-content shadow bg-base-300 rounded-box w-52 mt-4"
      tabindex="0"
    >
      <li>
        <a
          href="/#"
        >
          Profile
        </a>
      </li>
      <li>
        <a
          href="/#"
        >
          Settings
        </a>
      </li>
      <li>
        <a
          href="/#"
        >
          Logout
        </a>
      </li>
    </ul>
  </div>
</div>
`;
